--- monero/src/blockchain_db/lmdb/db_lmdb.h	2019-06-20 21:03:18.670817098 +0100
+++ haven/src/blockchain_db/lmdb/db_lmdb.h	2019-05-30 17:58:17.101592472 +0100
@@ -421,12 +405,6 @@
   // migrate from DB version 2 to 3
   void migrate_2_3();
 
-  // migrate from DB version 3 to 4
-  void migrate_3_4();
-
-  // migrate from DB version 4 to 5
-  void migrate_4_5();
-
   void cleanup_batch();
 
 private:
