--- monero/src/blockchain_db/lmdb/db_lmdb.cpp	2019-06-20 21:03:18.670817098 +0100
+++ haven/src/blockchain_db/lmdb/db_lmdb.cpp	2019-05-30 17:58:17.097592512 +0100
@@ -774,7 +704,7 @@
   bi.bi_diff_lo = (cumulative_difficulty & 0xffffffffffffffff).convert_to<uint64_t>();
   bi.bi_hash = blk_hash;
   bi.bi_cum_rct = num_rct_outs;
-  if (blk.major_version >= 4)
+  if (m_height > 0)
   {
     uint64_t last_height = m_height-1;
     MDB_val_set(h, last_height);
