--- monero/src/daemon/executor.cpp	2019-06-20 21:03:18.738816536 +0100
+++ haven/src/daemon/executor.cpp	2019-05-30 17:58:17.125592233 +0100
@@ -40,7 +40,7 @@
 
 namespace daemonize
 {
-  std::string const t_executor::NAME = "Monero Daemon";
+  std::string const t_executor::NAME = "Haven Daemon";
 
   void t_executor::init_options(
       boost::program_options::options_description & configurable_options
@@ -58,7 +58,7 @@
       boost::program_options::variables_map const & vm
     )
   {
-    LOG_PRINT_L0("Monero '" << MONERO_RELEASE_NAME << "' (v" << MONERO_VERSION_FULL << ") Daemonised");
-    return t_daemon{vm, public_rpc_port};
+    LOG_PRINT_L0("Haven '" << MONERO_RELEASE_NAME << "' (v" << MONERO_VERSION_FULL << ") Daemonised");
+    return t_daemon{vm};
   }
 
   bool t_executor::run_non_interactive(
