--- monero/src/cryptonote_core/blockchain.h	2019-06-20 21:03:18.734816570 +0100
+++ haven/src/cryptonote_core/blockchain.h	2019-05-30 17:58:17.125592233 +0100
@@ -57,6 +57,7 @@
 #include "checkpoints/checkpoints.h"
 #include "cryptonote_basic/hardfork.h"
 #include "blockchain_db/blockchain_db.h"
+#include "offshore/pricing_handler.h"
 
 namespace tools { class Notify; }
 
@@ -574,8 +570,8 @@
      * @brief get dynamic per kB or byte fee for a given block weight
      *
      * The dynamic fee is based on the block weight in a past window, and
-     * the current block reward. It is expressed by kB before v8, and
-     * per byte from v8.
+     * the current block reward. It is expressed by kB before v5, and
+     * per byte from v5.
      *
      * @param block_reward the current block reward
      * @param median_block_weight the median block weight in the past window
@@ -589,8 +585,8 @@
      * @brief get dynamic per kB or byte fee estimate for the next few blocks
      *
      * The dynamic fee is based on the block weight in a past window, and
-     * the current block reward. It is expressed by kB before v8, and
-     * per byte from v8.
+     * the current block reward. It is expressed by kB before v5, and
+     * per byte from v5.
      * This function calculates an estimate for a dynamic fee which will be
      * valid for the next grace_blocks
      *
@@ -1086,6 +1086,8 @@
 
     HardFork *m_hardfork;
 
+    offshore::PricingHandler *m_pricing_handler;
+
     network_type m_nettype;
     bool m_offline;
     difficulty_type m_fixed_difficulty;
