--- monero/src/cryptonote_core/CMakeLists.txt	2019-06-20 21:03:18.730816602 +0100
+++ haven/src/cryptonote_core/CMakeLists.txt	2019-05-30 17:58:17.125592233 +0100
@@ -1,4 +1,4 @@
-# Copyright (c) 2014-2019, The Monero Project
+# Copyright (c) 2014-2018, The Monero Project
 #
 # All rights reserved.
 #
@@ -30,7 +30,6 @@
   blockchain.cpp
   cryptonote_core.cpp
   tx_pool.cpp
-  tx_sanity_check.cpp
   cryptonote_tx_utils.cpp)
 
 set(cryptonote_core_headers)
@@ -40,7 +39,6 @@
   blockchain.h
   cryptonote_core.h
   tx_pool.h
-  tx_sanity_check.h
   cryptonote_tx_utils.h)
 
 monero_private_headers(cryptonote_core
