--- monero/src/cryptonote_core/tx_pool.cpp	2019-06-20 21:03:18.734816570 +0100
+++ haven/src/cryptonote_core/tx_pool.cpp	2019-05-30 17:58:17.125592233 +0100
@@ -82,8 +82,8 @@
 
     uint64_t get_transaction_weight_limit(uint8_t version)
     {
-      // from v8, limit a tx to 50% of the minimum block weight
-      if (version >= 8)
+      // from v5, limit a tx to 50% of the minimum block weight
+      if (version >= 5)
         return get_min_block_weight(version) / 2 - CRYPTONOTE_COINBASE_BLOB_RESERVED_SIZE;
       else
         return get_min_block_weight(version) - CRYPTONOTE_COINBASE_BLOB_RESERVED_SIZE;
