--- monero/src/wallet/message_transporter.h	2019-06-20 21:03:18.794816074 +0100
+++ haven/src/wallet/message_transporter.h	2019-06-05 12:45:14.769728847 +0100
@@ -48,7 +48,7 @@
   std::string source_transport_address;
   cryptonote::account_public_address destination_monero_address;
   std::string destination_transport_address;
-  crypto::chacha_iv iv;
+  crypto::chacha8_iv iv;
   crypto::public_key encryption_public_key;
   uint64_t timestamp;
   uint32_t type;
