--- monero/src/wallet/ringdb.h	2019-06-20 21:03:18.794816074 +0100
+++ haven/src/wallet/ringdb.h	2019-06-05 12:45:14.769728847 +0100
@@ -44,11 +44,11 @@
     void close();
     ~ringdb();
 
-    bool add_rings(const crypto::chacha_key &chacha_key, const cryptonote::transaction_prefix &tx);
-    bool remove_rings(const crypto::chacha_key &chacha_key, const std::vector<crypto::key_image> &key_images);
-    bool remove_rings(const crypto::chacha_key &chacha_key, const cryptonote::transaction_prefix &tx);
-    bool get_ring(const crypto::chacha_key &chacha_key, const crypto::key_image &key_image, std::vector<uint64_t> &outs);
-    bool set_ring(const crypto::chacha_key &chacha_key, const crypto::key_image &key_image, const std::vector<uint64_t> &outs, bool relative);
+    bool add_rings(const crypto::chacha8_key &chacha_key, const cryptonote::transaction_prefix &tx);
+    bool remove_rings(const crypto::chacha8_key &chacha_key, const std::vector<crypto::key_image> &key_images);
+    bool remove_rings(const crypto::chacha8_key &chacha_key, const cryptonote::transaction_prefix &tx);
+    bool get_ring(const crypto::chacha8_key &chacha_key, const crypto::key_image &key_image, std::vector<uint64_t> &outs);
+    bool set_ring(const crypto::chacha8_key &chacha_key, const crypto::key_image &key_image, const std::vector<uint64_t> &outs, bool relative);
 
     bool blackball(const std::pair<uint64_t, uint64_t> &output);
     bool blackball(const std::vector<std::pair<uint64_t, uint64_t>> &outputs);
