--- monero/src/device_trezor/trezor/messages_map.hpp	2019-06-20 21:03:18.754816404 +0100
+++ haven/src/device_trezor/trezor/messages_map.hpp	2019-05-30 17:58:17.129592192 +0100
@@ -1,4 +1,4 @@
-// Copyright (c) 2017-2019, The Monero Project
+// Copyright (c) 2017-2018, The Monero Project
 //
 // All rights reserved.
 //
@@ -62,14 +62,14 @@
     static messages::MessageType get_message_wire_number(const google::protobuf::Message & msg);
     static messages::MessageType get_message_wire_number(const std::string & msg_name);
 
-    template<class t_message=google::protobuf::Message>
+    template<class t_message>
     static messages::MessageType get_message_wire_number() {
       BOOST_STATIC_ASSERT(boost::is_base_of<google::protobuf::Message, t_message>::value);
       return get_message_wire_number(t_message::default_instance().GetDescriptor()->name());
     }
   };
 
-  template<class t_message=google::protobuf::Message>
+  template<class t_message>
   std::shared_ptr<t_message> message_ptr_retype(std::shared_ptr<google::protobuf::Message> & in){
     BOOST_STATIC_ASSERT(boost::is_base_of<google::protobuf::Message, t_message>::value);
     if (!in){
@@ -79,7 +79,7 @@
     return std::dynamic_pointer_cast<t_message>(in);
   }
 
-  template<class t_message=google::protobuf::Message>
+  template<class t_message>
   std::shared_ptr<t_message> message_ptr_retype_static(std::shared_ptr<google::protobuf::Message> & in){
     BOOST_STATIC_ASSERT(boost::is_base_of<google::protobuf::Message, t_message>::value);
     if (!in){
