--- monero/src/device/device_ledger.cpp	2019-06-20 21:03:18.742816503 +0100
+++ haven/src/device/device_ledger.cpp	2019-05-30 17:58:17.129592192 +0100
@@ -514,11 +470,11 @@
         return true;
     }
 
-    bool  device_ledger::generate_chacha_key(const cryptonote::account_keys &keys, crypto::chacha_key &key, uint64_t kdf_rounds) {
+    bool  device_ledger::generate_chacha_key(const cryptonote::account_keys &keys, crypto::chacha8_key &key, uint64_t kdf_rounds) {
         AUTO_LOCK_CMD();
 
         #ifdef DEBUG_HWDEVICE
-        crypto::chacha_key key_x;
+        crypto::chacha8_key key_x;
         cryptonote::account_keys keys_x = hw::ledger::decrypt(keys); 
         this->controle_device->generate_chacha_key(keys_x, key_x, kdf_rounds);
         #endif
@@ -527,10 +483,10 @@
 
         char prekey[200];
         memmove(prekey, &this->buffer_recv[0], 200);
-        crypto::generate_chacha_key_prehashed(&prekey[0], sizeof(prekey), key, kdf_rounds);
+	crypto::generate_chacha8_key(&prekey[0], sizeof(prekey), key);
 
         #ifdef DEBUG_HWDEVICE
-        hw::ledger::check32("generate_chacha_key_prehashed", "key", (char*)key_x.data(), (char*)key.data());
+        hw::ledger::check32("generate_chacha_key", "key", (char*)key_x.data(), (char*)key.data());
         #endif
 
       return true;
