--- monero/src/device/device_ledger.hpp	2019-06-20 21:03:18.742816503 +0100
+++ haven/src/device/device_ledger.hpp	2019-05-30 17:58:17.129592192 +0100
@@ -173,7 +155,7 @@
         /* ======================================================================= */
         bool  get_public_address(cryptonote::account_public_address &pubkey) override;
         bool  get_secret_keys(crypto::secret_key &viewkey , crypto::secret_key &spendkey) override;
-        bool  generate_chacha_key(const cryptonote::account_keys &keys, crypto::chacha_key &key, uint64_t kdf_rounds) override;
+        bool  generate_chacha_key(const cryptonote::account_keys &keys, crypto::chacha8_key &key, uint64_t kdf_rounds) override;
 
 
         /* ======================================================================= */
