--- monero/src/p2p/net_node.inl	2019-06-20 21:03:18.766816305 +0100
+++ haven/src/p2p/net_node.inl	2019-05-30 17:58:17.137592113 +0100
@@ -448,30 +437,21 @@
     std::set<std::string> full_addrs;
     if (nettype == cryptonote::TESTNET)
     {
-      full_addrs.insert("212.83.175.67:28080");
-      full_addrs.insert("5.9.100.248:28080");
-      full_addrs.insert("163.172.182.165:28080");
-      full_addrs.insert("195.154.123.123:28080");
-      full_addrs.insert("212.83.172.165:28080");
+      full_addrs.insert("seed01.testnet.havenprotocol.org:27749");
+      full_addrs.insert("seed02.testnet.havenprotocol.org:27749");
+      full_addrs.insert("seed03.testnet.havenprotocol.org:27749");
     }
     else if (nettype == cryptonote::STAGENET)
     {
-      full_addrs.insert("162.210.173.150:38080");
-      full_addrs.insert("162.210.173.151:38080");
     }
     else if (nettype == cryptonote::FAKECHAIN)
     {
     }
     else
     {
-      full_addrs.insert("107.152.130.98:18080");
-      full_addrs.insert("212.83.175.67:18080");
-      full_addrs.insert("5.9.100.248:18080");
-      full_addrs.insert("163.172.182.165:18080");
-      full_addrs.insert("161.67.132.39:18080");
-      full_addrs.insert("198.74.231.92:18080");
-      full_addrs.insert("195.154.123.123:18080");
-      full_addrs.insert("212.83.172.165:18080");
+      full_addrs.insert("seed01.mainnet.havenprotocol.org:17749");
+      full_addrs.insert("seed02.mainnet.havenprotocol.org:17749");
+      full_addrs.insert("seed03.mainnet.havenprotocol.org:17749");
     }
     return full_addrs;
   }
