--- monero/src/common/password.cpp	2019-06-20 21:03:18.718816701 +0100
+++ haven/src/common/password.cpp	2019-05-30 17:58:17.105592432 +0100
@@ -1,4 +1,4 @@
-// Copyright (c) 2014-2019, The Monero Project
+// Copyright (c) 2014-2018, The Monero Project
 // 
 // All rights reserved.
 // 
@@ -57,7 +57,7 @@
 
     DWORD mode_old;
     ::GetConsoleMode(h_cin, &mode_old);
-    DWORD mode_new = mode_old & ~((hide_input ? ENABLE_ECHO_INPUT : 0) | ENABLE_LINE_INPUT);
+    DWORD mode_new = mode_old & ~(hide_input ? ENABLE_ECHO_INPUT : 0);
     ::SetConsoleMode(h_cin, mode_new);
 
     bool r = true;
@@ -77,6 +77,10 @@
       }
       else if (ucs2_ch == L'\r')
       {
+        continue;
+      }
+      else if (ucs2_ch == L'\n')
+      {
         std::cout << std::endl;
         break;
       }
