--- monero/src/cryptonote_basic/cryptonote_format_utils.h	2019-06-20 21:03:18.730816602 +0100
+++ haven/src/cryptonote_basic/cryptonote_format_utils.h	2019-05-30 17:58:17.121592273 +0100
@@ -117,8 +117,7 @@
   bool calculate_block_hash(const block& b, crypto::hash& res, const blobdata *blob = NULL);
   bool get_block_hash(const block& b, crypto::hash& res);
   crypto::hash get_block_hash(const block& b);
-  bool get_block_longhash(const block& b, crypto::hash& res, uint64_t height);
-  crypto::hash get_block_longhash(const block& b, uint64_t height);
+  bool get_block_longhash(const block& b, cn_pow_hash_v3 &ctx, crypto::hash& res);
   bool parse_and_validate_block_from_blob(const blobdata& b_blob, block& b, crypto::hash *block_hash);
   bool parse_and_validate_block_from_blob(const blobdata& b_blob, block& b);
   bool parse_and_validate_block_from_blob(const blobdata& b_blob, block& b, crypto::hash &block_hash);
