--- monero/src/cryptonote_basic/difficulty.h	2019-06-20 21:03:18.730816602 +0100
+++ haven/src/cryptonote_basic/difficulty.h	2019-05-30 17:58:17.121592273 +0100
@@ -61,4 +51,5 @@
     difficulty_type next_difficulty(std::vector<std::uint64_t> timestamps, std::vector<difficulty_type> cumulative_difficulties, size_t target_seconds);
 
     std::string hex(difficulty_type v);
+    difficulty_type next_difficulty_v2(std::vector<std::uint64_t> timestamps, std::vector<difficulty_type> cumulative_difficulties, size_t target_seconds);
 }
